src/WinRT.Runtime/Marshalers.cs

```diff
@@ -25,55 +26,105 @@ public static void Dispose(this GCHandle handle)
+        {
+            IntPtr reserved1;
+            int reserved2;
+            int reserved3;
+            int reserved4;
+            int reserved5;
+        }
+
+        private HSTRING_HEADER _header;
+        private GCHandle _gchandle;
+
+        public struct Pinnable
+        {
+            private HSTRING_HEADER _header;
+            private string _value;
+
+            public Pinnable(string value)
+            {
+                _value = value ?? "";
+                _header = default;
+            }
+
+            public ref readonly char GetPinnableReference()
+            {
+                return ref _value.GetPinnableReference();
+            }
+
+            public unsafe IntPtr GetAbi()
+            {
+                if (_value == "")
+                {
+                    return IntPtr.Zero;
+                }
+                IntPtr hstring;
+                Marshal.ThrowExceptionForHR(Platform.WindowsCreateStringReference(
+                    (char*)Unsafe.AsPointer(ref Unsafe.AsRef(in GetPinnableReference())),
+                    _value.Length,
+                    (IntPtr*)Unsafe.AsPointer(ref _header),
+                    &hstring));
+                return hstring;
+            }
+        }
+
+        public static Pinnable CreatePinnable(string value) => new(value);
+
+        public static IntPtr GetAbi(Pinnable p) => p.GetAbi();
+
+        public MarshalString(string value)
+        {
+            _gchandle = GCHandle.Alloc(value, GCHandleType.Pinned);
+        }
 
         public void Dispose()
         {
             _gchandle.Dispose();
-            Marshal.FreeHGlobal(_header);
-            _header = IntPtr.Zero;
         }
 
-        public static unsafe MarshalString CreateMarshaler(string value)
+        public static MarshalString CreateMarshaler(string value)
         {
-            if (value == null) return null;
-
-            var m = new MarshalString();
+            return string.IsNullOrEmpty(value) ? null : new MarshalString(value);
+        }
 
-            bool success = false;
+        public unsafe IntPtr GetAbi()
+        {
+            static bool Dispose(MarshalString m) { m.Dispose(); return false; };
             try
             {
-                m._gchandle = GCHandle.Alloc(value, GCHandleType.Pinned);
-                m._header = Marshal.AllocHGlobal(24); // sizeof(HSTRING_HEADER)
-                fixed (void* chars = value, handle = &m._handle)
-                {
+                var value = (string)_gchandle.Target;
+                fixed (char* chars = value)
+                { 
+                    IntPtr hstring;
                     Marshal.ThrowExceptionForHR(Platform.WindowsCreateStringReference(
-                        (char*)chars, value.Length, (IntPtr*)m._header, (IntPtr*)handle));
-                };
-                success = true;
-                return m;
+                        chars, value.Length, (IntPtr*)Unsafe.AsPointer(ref _header), &hstring));

```

Comment:
What guarantees that `_header` is pinned here? You can use `Unsafe.AsPointer` only if the item is pinned by some other means.

Reply:
@jkotas, this was taken from your example.  _header is a struct member of the Pinnable struct.  assuming we pass by ref as you point out above, all instances are on the stack, yes?

Reply:
> _header is a struct member of the Pinnable struct

If I am reading the code correctly, `_header` is a field on MarshalString class here. I do not see what would keep MarshalString pinned.

Reply:
sorry, I was looking at the wrong code.  Oh, this is unfortunate - I was hoping to eliminate allocating the header on the heap, but now I don't see a way around that.